platform/pulpcore/app/models/storage.py

```diff
@@ -205,13 +117,12 @@ def _save(path, content):
             str: Final storage path.
         """
         os.makedirs(os.path.dirname(path), exist_ok=True)
-
         try:
             try:
-                os.link(content.name, path)
+                os.link(content.file.name, path)
             except OSError as e:
                 if e.errno in (errno.EXDEV, errno.ENOENT):
-                    FileSystem.copy(path, content)
+                    super()._save(path, content)
```

Comment:
Most (perhaps all) of the unwanted behavior is in FileSystemStorage._save().
- directory creation is not thread safe.
- new file-name is created when file already exists.
 - uses os.rename() instead of link.

It also tries to create the dir and rename the file which is 2 unnecessary system calls.

However in this case, it's only being called when the directory has already been created and the link() would raise `FileExistsError`.  So it works for now.  What if other undesirable behaviors get introduced in `FIleSystemStorage._save()`?  I don't think removing `FileSystem.copy()` and saving 11 lines of code is worth the risk of using `FIleSystemStorage._save()`.

Reply:
Thank you for the feedback. I read through the _save() method in Django and  it looks like it already does what we want it to do for TemporaryUploadedFile objects. When Django sees that the file object has a temporary_file_path method, it attempts to move the file instead of copying it. 

Why is the directory creation not thread safe? It looks like it tries to create the repo and if it already exists it keeps going

Instead of implementing the _save() method we can implement save() and get_available_name(). The get_available_name() method should raise an Exception when the file already exists. save() can catch that exception and just keep going without doing any other operations.

I am not sure that I understand why os.rename is not desired for our use case. Could you please elaborate?